File: Amazon_LeetCode_Python_C++/ArrayString/20_valid_parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        if (s.size() == 0) {
            return true;
        }
        if (s.size() == 1) {
            return false;
        }
        for (int i = 0; i < s.size(); i++) {
            if (s.at(i) == '{' || s.at(i) == '[' || s.at(i) == '(') {
                stack.push(s.at(i));
            }
            else {
                if (stack.empty()) {
                    return false;
                }
                if ((s.at(i) == '}' && stack.top() != '{') || 
                    (s.at(i) == ']' && stack.top() != '[') ||
                    (s.at(i) == ')' && stack.top() != '(')) {
                    return false;
                }
                stack.pop();
            }
        }
        return stack.empty();
    }
};

File: Amazon_LeetCode_Python_C++/ArrayString/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        mp = {')':'(', ']':'[', '}':'{'};
        sp = []
        
        for i, c in enumerate(s):
            if c not in mp:
                sp.append(c)
            else:
                if len(sp) == 0 or mp[c] != sp[-1]:
                    return False
                sp.pop()
                
        return len(sp) == 0
